File: api/ccs/ccs_solve_tril.m
function b = ccs_solve_tril(L, b)
% ccs_solve_tril Solves L\b, where L is lower triangular
%     b = ccs_solve_tril(L, b)
%  The matrix is assumed to be lower triangular. The right-hand side vector
%  will be overwritten by L\b at output
%
% See also ccs_solve_triu, ccs_solve_trilt

%#codegen -args {ccs_matrix, coder.typeof(0, [inf, 1])}

coder.inline('always');

for j = 1:int32(length(L.col_ptr)) - 1
    start = L.col_ptr(j);
    assert(L.row_ind(start) == j && L.val(start) ~= 0);

    b(j) = b(j) / L.val(start);

    for k = (start + 1):(L.col_ptr(j+1) - 1)
        i = L.row_ind(k);
        % This can be done concurrently
        b(i) = b(i) - L.val(k) * b(j);
    end
end

function test %#ok<DEFNU>
%!test
%! for k=1:100
%!     T = tril(rand(10,10)); b = rand(10, 1);
%!     if condest(T) < 1.e5
%!         L = ccs_createFromSparse(sparse(T));
%!         us = ccs_solve_tril(L, b);
%!         assert(norm(T * us - b) <= 1.e-10);
%!     end
%! end

File: api/ccs/codegen/lib/ccs_solve_tril/ccs_solve_tril.c
#include "ccs_solve_tril.h"
#include "ccs_solve_tril_types.h"

void ccs_solve_tril(const CCS_Matrix *L, emxArray_real_T *b)
{
  int i;
  int i1;
  int i2;
  int i3;
  int j;
  int k;
  i = L->col_ptr->size[0];
  for (j = 0; j <= i - 2; j++) {
    b->data[j] /= L->val->data[L->col_ptr->data[j] - 1];
    i1 = L->col_ptr->data[j] + 1;
    i2 = L->col_ptr->data[j + 1] - 1;
    for (k = i1; k <= i2; k++) {
      i3 = L->row_ind->data[k - 1] - 1;
      b->data[i3] -= L->val->data[k - 1] * b->data[j];
    }
  }
}

void ccs_solve_tril_initialize(void)
{
}

void ccs_solve_tril_terminate(void)
{
}
